R/gm-auth.R

```diff
@@ -0,0 +1,207 @@
+#' imply that subsequent requests are made with an API key in lieu of a token.
+#'
+#' @family auth functions
+#' @export
+#' @examples
+#' \dontrun{
+#' gm_deauth()
+#' }
+gm_deauth <- function() {
+  .auth$clear_cred()
+  invisible()
+}
+
+#' Produce configured token
+#'
+#' @eval gargle:::PREFIX_token_description(gargle_lookup_table)
+#' @eval gargle:::PREFIX_token_return()
+#'
+#' @family low-level API functions
+#' @export
+#' @examples
+#' \dontrun{
+#' gm_token()
+#' }
+gm_token <- function() {
+  if (!gm_has_token()) {
+    gm_auth()
+  }
+  httr::config(token = .auth$cred)
+}
+
+#' Is there a token on hand?
+#'
+#' Reports whether gmailr has stored a token, ready for use in downstream
+#' requests.
+#'
+#' @return Logical.
+#' @export
+#'
+#' @examples
+#' gm_has_token()
+gm_has_token <- function() {
+  inherits(.auth$cred, "Token2.0")
+}
+
+#' Edit auth configuration
+#'
+#' @description
+#' These functions give the user more control over auth than what is possible
+#' with [gm_auth()]. `gm_auth_configure()` gives control of:
+#'   * The OAuth app, which is used when obtaining a user token.
+#'
+#' See the vignette [How to get your own API
+#' credentials](https://gargle.r-lib.org/articles/get-api-credentials.html) for
+#' more.
+#'
+#' @param app OAuth app.
+#' @inheritParams gargle::oauth_app_from_json
+#'
+#' @return
+#'   * `gm_auth_configure()`: An object of R6 class [gargle::AuthState],
+#'     invisibly.
+#'   * `gm_oauth_app()`: the current user-configured [httr::oauth_app()].
+#'
+#' @family auth functions
+#' @export
+#' @aliases gm_auth_config
+#' @examples
+#' # see the current user-configured OAuth app (probaby `NULL`)
+#' gm_oauth_app()
+#'
+#' if (require(httr)) {
+#'
+#'   # store current state, so we can restore
+#'   original_app <- gm_oauth_app()
+#'
+#'   # bring your own app via client id (aka key) and secret
+#'   google_app <- httr::oauth_app(
+#'     "my-awesome-google-api-wrapping-package",
+#'     key = "123456789.apps.googleusercontent.com",
+#'     secret = "abcdefghijklmnopqrstuvwxyz"
+#'   )
+#'   gm_auth_configure(app = google_app)
+#'
+#'   # confirm current app
+#'   gm_oauth_app()
+#'
+#'   # restore original state
+#'   gm_auth_configure(app = original_app)
+#'   gm_oauth_app()
+#' }
+#'
+#' \dontrun{
+#' # bring your own app via JSON downloaded from Google Developers Console
+#' gm_auth_configure(
+#'   path = "/path/to/the/JSON/you/downloaded/from/google/dev/console.json"
+#' )
+#' }
+#'
+gm_auth_configure <- function(app, path) {
+  if (!xor(missing(app), missing(path))) {
+    stop("Must supply exactly one of `app` and `path`", call. = FALSE)
+  }
+  if (!missing(path)) {
+    stopifnot(is_string(path))
+    app <- gargle::oauth_app_from_json(path)
+  }
+  stopifnot(is.null(app) || inherits(app, "oauth_app"))
+
+  .auth$set_app(app)
+
+  invisible(.auth)
+}
+
+#' @export
+#' @rdname gm_auth_configure
+gm_oauth_app <- function() .auth$app
```

Comment:
Since gmailr is strictly BYOA (bring your own app), I suspect this should have some logic that forces the user to confront that, at the first call.

I mean, I'm sure this fails now, but the error message probably isn't terribly helpful.